client/fingerprint/dynamic_host_volumes.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package fingerprint
+
+import (
+	"context"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	hvm "github.com/hashicorp/nomad/client/hostvolumemanager"
+	"github.com/hashicorp/nomad/helper"
+)
+
+func NewPluginsHostVolumeFingerprint(logger hclog.Logger) Fingerprint {
+	return &DynamicHostVolumePluginFingerprint{
+		logger: logger.Named("host_volume_plugins"),
+	}
+}
+
+var _ ReloadableFingerprint = &DynamicHostVolumePluginFingerprint{}
+
+type DynamicHostVolumePluginFingerprint struct {
+	logger hclog.Logger
+}
+
+func (h *DynamicHostVolumePluginFingerprint) Reload() {
+	// host volume plugins are re-detected on agent reload
+}
+
+func (h *DynamicHostVolumePluginFingerprint) Fingerprint(request *FingerprintRequest, response *FingerprintResponse) error {
+	// always add "mkdir" plugin
+	h.logger.Debug("detected plugin built-in", "plugin_id", "mkdir", "version", hvm.HostVolumePluginMkdirVersion)
+	defer response.AddAttribute("plugins.dhv.version.mkdir", hvm.HostVolumePluginMkdirVersion)
+	response.Detected = true
+
+	// this config value will be empty in -dev mode
+	pluginDir := request.Config.DynamicHostVolumePluginPath
+	if pluginDir == "" {
+		return nil
+	}
+
+	plugins, err := GetHostVolumePluginVersions(h.logger, pluginDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			h.logger.Debug("plugin dir does not exist", "dir", pluginDir)
+		} else {
+			h.logger.Warn("error finding plugins", "dir", pluginDir, "error", err)
+		}
+		return nil // don't halt agent start
+	}
+
+	// if this was a reload, wipe what was there before
+	for k := range request.Node.Attributes {
+		if strings.HasPrefix(k, "plugins.dhv.") {
+			response.RemoveAttribute(k)
+		}
+	}
+
+	// set the attribute(s)
+	for plugin, version := range plugins {
+		h.logger.Debug("detected plugin", "plugin_id", plugin, "version", version)
+		response.AddAttribute("plugins.dhv.version."+plugin, version)
```

Comment:
I could maybe be convinced otherwise on the plugin configuration field I noted below, but here if we have the `dhv` abbreviation it loses all context by the time it gets to be user-visible. We end up with a mysterious acronym in the middle of a very long list of attributes.

```suggestion
		response.AddAttribute("plugins.host_volume.version."+plugin, version)
```